File: packages/contracts/python/src/oocc_contracts/validators.py
"""Runtime validators for the OOCC trace and viz-plan contracts.

Each validator checks a raw dict against the canonical JSON Schema — which
enforces conditional rules the generated Pydantic models don't express on
their own (e.g. `error` is required when `status` is runtime_error or
compile_error; `returned` is present iff `event == "return"`) — and then
parses it into the corresponding Pydantic model for typed downstream use.
"""

from __future__ import annotations

import json
from functools import cache
from importlib import resources
from typing import Any

from jsonschema import Draft202012Validator

from oocc_contracts.generated.analysis_model import Analysis
from oocc_contracts.generated.trace_model import Trace
from oocc_contracts.generated.viz_plan_model import VizPlan

__all__ = [
    "ContractValidationError",
    "validate_analysis",
    "validate_trace",
    "validate_viz_plan",
]


class ContractValidationError(ValueError):
    """Raised when a payload does not conform to a contract schema."""

    def __init__(self, schema_filename: str, errors: list[str]) -> None:
        self.schema_filename = schema_filename
        self.errors = errors
        joined = "\n".join(f"  - {e}" for e in errors)
        super().__init__(f"{schema_filename} failed schema validation:\n{joined}")


@cache
def _load_validator(schema_filename: str) -> Draft202012Validator:
    schema_text = resources.files("oocc_contracts.generated").joinpath(schema_filename).read_text()
    schema = json.loads(schema_text)
    Draft202012Validator.check_schema(schema)
    return Draft202012Validator(schema)


def _validate(schema_filename: str, data: dict[str, Any]) -> None:
    validator = _load_validator(schema_filename)
    errors = sorted(validator.iter_errors(data), key=lambda e: list(e.absolute_path))
    if errors:
        messages = [f"{list(e.absolute_path)}: {e.message}" for e in errors]
        raise ContractValidationError(schema_filename, messages)


def validate_trace(data: dict[str, Any]) -> Trace:
    """Validate a raw trace payload against trace.schema.json and parse it.

    Raises ContractValidationError on any schema violation.
    """
    _validate("trace.schema.json", data)
    return Trace.model_validate(data)


def validate_viz_plan(data: dict[str, Any]) -> VizPlan:
    """Validate a raw viz-plan payload against viz-plan.schema.json and parse it.

    Raises ContractValidationError on any schema violation.
    """
    _validate("viz-plan.schema.json", data)
    return VizPlan.model_validate(data)


def validate_analysis(data: dict[str, Any]) -> Analysis:
    """Validate a raw analysis payload against analysis.schema.json and parse it.

    Raises ContractValidationError on any schema violation.
    """
    _validate("analysis.schema.json", data)
    return Analysis.model_validate(data)

File: packages/contracts/ts/src/validators.ts
import Ajv2020 from "ajv/dist/2020.js";
import addFormats from "ajv-formats";
import type { Analysis } from "./generated/analysis";
import analysisSchema from "./generated/analysis.schema.json";
import type { Trace } from "./generated/trace";
import traceSchema from "./generated/trace.schema.json";
import type { VizPlan } from "./generated/viz_plan";
import vizPlanSchema from "./generated/viz-plan.schema.json";

// `strictRequired` is disabled because our `if`/`then` conditionals (e.g.
// "error" required when status is runtime_error/compile_error) reference a
// property declared on the parent schema rather than restated inside the
// `then` branch — valid JSON Schema (and this repo's Python validator,
// backed by the `jsonschema` package, accepts it under the same rules), but
// Ajv's strict mode flags it as an unusual pattern. Every other strict check
// stays on.
const ajv = new Ajv2020({ allErrors: true, strict: true, strictRequired: false });
addFormats(ajv);

const validateTraceSchema = ajv.compile<Trace>(traceSchema);
const validateVizPlanSchema = ajv.compile<VizPlan>(vizPlanSchema);
const validateAnalysisSchema = ajv.compile<Analysis>(analysisSchema);

export class ContractValidationError extends Error {
  readonly schemaFilename: string;
  readonly errors: string[];

  constructor(schemaFilename: string, errors: string[]) {
    super(`${schemaFilename} failed schema validation:\n${errors.map((e) => `  - ${e}`).join("\n")}`);
    this.name = "ContractValidationError";
    this.schemaFilename = schemaFilename;
    this.errors = errors;
  }
}

/**
 * Validate a raw trace payload against trace.schema.json.
 *
 * Throws ContractValidationError on any schema violation, including the
 * conditional rules (e.g. `returned` present iff `event === "return"`) that
 * the generated Trace type alone does not enforce at compile time.
 */
export function validateTrace(data: unknown): Trace {
  if (!validateTraceSchema(data)) {
    throw new ContractValidationError(
      "trace.schema.json",
      (validateTraceSchema.errors ?? []).map((e) => `${e.instancePath || "/"} ${e.message ?? ""}`.trim()),
    );
  }
  return data;
}

/** Validate a raw viz-plan payload against viz-plan.schema.json. */
export function validateVizPlan(data: unknown): VizPlan {
  if (!validateVizPlanSchema(data)) {
    throw new ContractValidationError(
      "viz-plan.schema.json",
      (validateVizPlanSchema.errors ?? []).map((e) => `${e.instancePath || "/"} ${e.message ?? ""}`.trim()),
    );
  }
  return data;
}

/** Validate a raw analysis payload against analysis.schema.json. */
export function validateAnalysis(data: unknown): Analysis {
  if (!validateAnalysisSchema(data)) {
    throw new ContractValidationError(
      "analysis.schema.json",
      (validateAnalysisSchema.errors ?? []).map((e) => `${e.instancePath || "/"} ${e.message ?? ""}`.trim()),
    );
  }
  return data;
}
